docs/api_overview.md

```diff
@@ -0,0 +1,472 @@
+# Proxy-Wasm C++ SDK API Overview
+
+This page explains the main concepts and structure of the C++ SDK API. For
+detailed API semantics, refer to the doc comments in the header files listed in
+the [codemap](#codemap) section, in particular [proxy_wasm_api.h].
+
+## Concepts and terminology
+
+* **Plugin**: a unit of code for customizing network proxy
+  functionality. Plugins can be written in a variety of programming languages
+  (in this case, C++), and are compiled to Wasm modules.
+  
+* **Wasm VM**: the virtual machine that runs plugins. A network proxy may
+  internally instantiate multiple Wasm VMs to run plugins, for example one per
+  worker thread. A given Wasm VM can host multiple plugins at the same time.
+  
+* **Hostcalls and callbacks**: *callbacks* are the entry points to the plugin:
+  they are functions defined by the plugin that the proxy invokes when events of
+  interest (e.g. plugin startup, receipt of a new incoming HTTP request)
+  occur. *Hostcalls* are function calls that the plugin can make to the host
+  (the network proxy) to query additional state or perform actions, such as
+  reading and modifying HTTP response headers.
+  
+* **Contexts**: objects that are associated with the current plugin, or the
+  current "stream" (either HTTP/gRPC request or, if L4 proxying, TCP stream). In
+  the C++ SDK, Contexts serve multiple purposes:
+  
+  * They are the receivers of Proxy-Wasm callbacks: for example, when HTTP
+    request headers are received, that is represented as a call to the
+    `onRequestHeaders` method of a Context object. Callbacks are declared as
+    virtual methods on a `Context` base class, which plugin code can override.
+  * They provide the interface for most hostcalls, particularly those associated
+    with the current stream. The hostcalls are defined as methods on `Context`
+    base classes, which plugin code can call.
+    
+  There are two types of Contexts: *Root Contexts* and *Stream Contexts*
```

Comment:
"Root Context" is a leftover from forever-ago, I'd change it to "Plugin Context (also known as Root Context)" to match the spec docs.